### PR TITLE
[consensus] support a recovery mode if consensusdb is gone

### PR DESCRIPTION
**Pull request by zekun000**
This commit allows node to sync without a consensus db and the process would exit after it finishes and the validator can operator normally after restart.

It's not ideal that it requires another restart but it's better than what we have today.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4475)

### CODE REVIEW

File: consensus/src/recovery_manager.rs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                            )
+                        }
+                        unexpected_event => unreachable!("Unexpected event: {:?}", unexpected_event),
+                    }
+                    .with_context(|| format!("from peer {}", peer_id));
+
+                    match result {
+                        Ok(_) => {
+                            info!("Recovery finishes for epoch {}, RecoveryManager stopped. Please restart the node", self.epoch_state.epoch);
+                            process::exit(0);
```
Comment (igor-aptos) [inline consensus/src/recovery_manager.rs:133]: any reason we cannot start round manager after this finishes?

File: consensus/src/recovery_manager.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                        VerifiedEvent::VoteMsg(vote_msg) => {
+                            monitor!("process_recovery", self.process_vote_msg(*vote_msg).await)
+                        }
+                        VerifiedEvent::UnverifiedSyncInfo(sync_info) => {
+                            monitor!(
+                                "process_recovery",
+                                self.sync_up(&sync_info, peer_id).await
+                            )
+                        }
+                        unexpected_event => unreachable!("Unexpected event: {:?}", unexpected_event),
```
Comment (igor-aptos) [inline consensus/src/recovery_manager.rs:126]: I might be missing something, but what guarantees we don't receive a different message? 

especially if a node crashes and restarts in this mode, other nodes might have messages to send in response to messages from this node before the crash?

**Review by igor-aptos — COMMENTED** [review]
overall looks good, and too simple :)

File: consensus/src/recovery_manager.rs (hunk 1 above)
Comment (zekun000) [reply consensus/src/recovery_manager.rs:133]: it's a little messy to wire it back to epoch manager, and I don't want to do the major refactor at this point of time

File: consensus/src/recovery_manager.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/recovery_manager.rs:126]: it's not controlled by other nodes, it's an internal channel receives messages from epoch manager.

File: consensus/src/recovery_manager.rs (hunk 1 above)
Comment (sasha8) [inline consensus/src/recovery_manager.rs:133]: Do we have to restart the node? By the time we restart we might need to sync again. 
Cannot we go back to the epoch manger and create the round manager without restarting?

File: consensus/src/recovery_manager.rs (hunk 2 above)
Comment (sasha8) [reply consensus/src/recovery_manager.rs:126]: I just had the same comment.
What if the epoch manager gets a commit message and then forward it here?

File: consensus/src/recovery_manager.rs (hunk 1 above)
Comment (sasha8) [reply consensus/src/recovery_manager.rs:133]: What if by the time we restart we need to sync again?

Comment (zekun000) [reply consensus/src/recovery_manager.rs:133]: after restart it'll sync through round manager

File: consensus/src/recovery_manager.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    pub async fn start(
+        mut self,
+        mut event_rx: aptos_channel::Receiver<
+            (Author, Discriminant<VerifiedEvent>),
+            (Author, VerifiedEvent),
+        >,
+        close_rx: oneshot::Receiver<oneshot::Sender<()>>,
+    ) {
+        info!(epoch = self.epoch_state.epoch, "RecoveryManager started");
+        let mut close_rx = close_rx.into_stream();
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/recovery_manager.rs:106]: don't remember, it's copy paste

File: consensus/src/persistent_liveness_storage.rs [hunk 4]
```diff
@@ -403,11 +408,11 @@ impl PersistentLivenessStorage for StorageWriteProxy {
                     initial_data.highest_2chain_timeout_certificate().as_ref().map_or("None".to_string(), |v| v.to_string()),
                 );
 
-                LivenessStorageData::RecoveryData(initial_data)
+                LivenessStorageData::FullRecoveryData(initial_data)
             }
             Err(e) => {
                 error!(error = ?e, "Failed to construct recovery data");
-                LivenessStorageData::LedgerRecoveryData(ledger_recovery_data)
+                LivenessStorageData::PartialRecoveryData(ledger_recovery_data)
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/persistent_liveness_storage.rs:415]: we need that to construct the RecoveryData

File: consensus/src/recovery_manager.rs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        self.sync_up(sync_info, author).await
+    }
+
+    pub async fn sync_up(&mut self, sync_info: &SyncInfo, peer: Author) -> Result<RecoveryData> {
+        sync_info.verify(&self.epoch_state.verifier)?;
+        ensure!(
+            sync_info.highest_round() > self.last_committed_round,
+            "[RecoveryManager] Received sync info has lower round number than committed block"
+        );
+        ensure!(
```
Comment (sitalkedia) [inline consensus/src/recovery_manager.rs:73]: How is epoch change handled in this mode? I don't see epoch manager doing anything special for recovery mode.

File: consensus/src/epoch_manager.rs [hunk 6]
```diff
@@ -680,16 +701,19 @@ impl EpochManager {
...
+            LivenessStorageData::FullRecoveryData(initial_data) => {
+                self.start_round_manager(
+                    initial_data,
+                    epoch_state,
+                    onchain_config.unwrap_or_default(),
+                )
+                .await
+            }
+            LivenessStorageData::PartialRecoveryData(ledger_data) => {
+                self.start_recovery_manager(ledger_data, epoch_state).await
```
Comment (sitalkedia) [inline consensus/src/epoch_manager.rs:714]: It would be helpful to add a log message here with something like - validator is starting up in recovery mode and will restart once recovery finishes.

File: consensus/src/epoch_manager.rs [hunk 7]
```diff
@@ -542,6 +534,35 @@ impl EpochManager {
...
+        ledger_data: LedgerRecoveryData,
+        epoch_state: EpochState,
+    ) {
+        let network_sender = NetworkSender::new(
+            self.author,
+            self.network_sender.clone(),
+            self.self_sender.clone(),
+            epoch_state.verifier.clone(),
+        );
+        let (round_manager_tx, round_manager_rx) = aptos_channel::new(
```
Comment (sitalkedia) [inline consensus/src/epoch_manager.rs:548]: `round_manager` -> `recovery_manager`

File: consensus/src/epoch_manager.rs [hunk 8]
```diff
@@ -542,6 +534,35 @@ impl EpochManager {
...
+            QueueStyle::LIFO,
+            1,
+            Some(&counters::ROUND_MANAGER_CHANNEL_MSGS),
```
Comment (sitalkedia) [inline consensus/src/epoch_manager.rs:551]: `ROUND_MANAGER_CHANNEL_MSGS` -> `RECOVERY_MANAGER_CHANNEL_MSGS`

File: consensus/src/epoch_manager.rs [hunk 9]
```diff
@@ -542,6 +534,35 @@ impl EpochManager {
...
+        );
+        self.round_manager_tx = Some(round_manager_tx);
```
Comment (sitalkedia) [inline consensus/src/epoch_manager.rs:553]: Let's rename this to `round_or_recovery_manager_tx` or something like that to avoid confusion.

File: consensus/src/recovery_manager.rs (hunk 5 above)
Comment (sitalkedia) [reply consensus/src/recovery_manager.rs:73]: Never mind, figured out we are creating a new recovery manager in case of epoch change.

File: testsuite/smoke-test/src/consensus/consensusdb_recovery.rs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+};
+use consensus::CONSENSUS_DB_NAME;
+use forge::{HealthCheckError, NodeExt, Swarm};
+use std::{
+    fs,
+    time::{Duration, Instant},
+};
+
+#[tokio::test]
+async fn test_consensusdb_recovery() {
```
Comment (sitalkedia) [inline testsuite/smoke-test/src/consensus/consensusdb_recovery.rs:16]: Can you add a test that ensures epoch change? May be just shorten the epoch length ane ensure we wait for some time for an epoch change?

File: consensus/src/recovery_manager.rs (hunk 5 above)
Comment (zekun000) [reply consensus/src/recovery_manager.rs:73]: epoch manager is going to shutdown the recovery manager like round manager and start new epoch normally

File: consensus/src/epoch_manager.rs (hunk 6 above)
Comment (zekun000) [reply consensus/src/epoch_manager.rs:714]: there's a log message when start the recovery manager

File: consensus/src/epoch_manager.rs (hunk 8 above)
Comment (zekun000) [reply consensus/src/epoch_manager.rs:551]: it shared the same channel as round manager, I don't think it's worth a separate counter

File: testsuite/smoke-test/src/consensus/consensusdb_recovery.rs (hunk 10 above)
Comment (zekun000) [reply testsuite/smoke-test/src/consensus/consensusdb_recovery.rs:16]: epoch change won't test this path, state sync would catch up to the new epoch and start with round manager

**Review by igor-aptos — APPROVED** [review]

File: testsuite/smoke-test/src/consensus/consensusdb_recovery.rs (hunk 10 above)
Comment (sitalkedia) [reply testsuite/smoke-test/src/consensus/consensusdb_recovery.rs:16]: Correct me if I am missing something but its possible that the recovery manager thread is doing state sync, while the epoch manager receives an epoch chaging proof, thus terminating the recreating a new recovery manager. Do you think there is a way to test that code path? I agree it should be a corner case though.

Comment (zekun000) [reply testsuite/smoke-test/src/consensus/consensusdb_recovery.rs:16]: it can not terminating an existing state sync, the close request needs to wait until the state sync finishes

**Review by sitalkedia — APPROVED** [review]
